File: s3-plugin-server/src/main/java/com/gu/teamcity/S3ConfigController.java
package com.gu.teamcity;

import jetbrains.buildServer.controllers.MultipartFormController;
import jetbrains.buildServer.web.openapi.WebControllerManager;
import org.jetbrains.annotations.NotNull;
import org.springframework.web.servlet.ModelAndView;
import org.springframework.web.servlet.view.RedirectView;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

public class S3ConfigController extends MultipartFormController {
    private S3Extension extension;

    public S3ConfigController(@NotNull S3Extension extension, @NotNull WebControllerManager webControllerManager) {
        this.extension = extension;
        webControllerManager.registerController("/app/s3/**", this);
    }

    @Override
    protected ModelAndView doPost(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse) {
        String bucket = httpServletRequest.getParameter("bucketName");
        extension.bucketName = bucket;

        return new ModelAndView(new RedirectView("/admin/admin.html?item=S3"));
    }
}

File: s3-plugin-server/src/main/scala/com/gu/teamcity/S3ConfigController.scala
package com.gu.teamcity

import jetbrains.buildServer.controllers.MultipartFormController
import jetbrains.buildServer.web.openapi.WebControllerManager
import org.jetbrains.annotations.NotNull
import org.springframework.web.servlet.ModelAndView
import org.springframework.web.servlet.view.RedirectView
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

class S3ConfigController(extension: S3ConfigManager, webControllerManager: WebControllerManager) extends MultipartFormController {
  webControllerManager.registerController("/app/s3/**", this)

  protected def doPost(request: HttpServletRequest, response: HttpServletResponse): ModelAndView = {
    def param(name: String) = S3ConfigController.emptyAsNone(request.getParameter(name))

    extension.update(S3Config(
      param("artifactBucket"), param("buildManifestBucket"), param("tagManifestBucket"),
      param("accessKey"), param("secretKey")
    ))

    new ModelAndView(new RedirectView("/admin/admin.html?item=S3"))
  }
}

object S3ConfigController {
  def emptyAsNone(s: String): Option[String] = Option(s).filterNot(_.trim.isEmpty)
}